### PR TITLE
Profiling off fix

### PR DESCRIPTION
**Pull request by DSMishler**
A response to https://github.com/ICLDisco/parsec/issues/529.

Applied the patch suggested by @bosilca, but there were two more things that needed done.

1. Sometimes the `profiling_array` is not initialized. This was fixed to now have `pins.c` check first in `parsec_pins_instrument`.
2. The `profiling_array` wasn't truly being initialized to `-1`. Saying `static int myarr[200] = {-1}` only sets the first element to `-1` with the current PaRSEC compiler settings. I added code to truly initialize all tasks.

I will continue a few tests of my own and I encourage you to double-check as well. I'll let you know if I find anything else!

### CODE REVIEW

**Comment by DSMishler** [comment]
I'm doing some testing in this PR and I'm noticing some issues. Don't merge yet until these are resolved.

File: parsec/interfaces/ptg/ptg-compiler/jdf2c.c
Comment (therault) [inline parsec/interfaces/ptg/ptg-compiler/jdf2c.c:1602]: We shouldn't need `%s_profiling_array_counter_limit` in this structure: AFAIU, it's always `2*nb_task_classes`.

Comment (bosilca) [reply parsec/interfaces/ptg/ptg-compiler/jdf2c.c:1602]: or 4 ... depending if we profile the internal startup tasks.

**Comment by bosilca** [comment]
I have quite a few comments, because this make little sense to me.

1. the profiling_array of each taskpool is set during the taskpool constructor to point to the statically allocated write_check_profiling_array. This initialization is protected by the profile #define (which should always be consistently set across all builds from the same runtime), and by the `write_check_profiling_array[0] == -1`.
2. Because of this test it seems clear to me that we were not trying to initialize all the array but only the first element. If that element was not zero it simply means that the array should contains the correct entries in the profiling dictionary.
3. Instead of doing a loop and setting all values to -1 you should check what tasks have the `profile=off` in the constructor and only set their dictionary entries to -1. That would also remove the need for storing the number of entries in the profiling array.

**Comment by therault** [comment]
@DSMishler we had a discussion about this PR during the PR reviews today.

 - Please leave the static initialization as is ( `profiling_array[...] = {-1, },` in the generated code ), this is used by other features. The -1 will have to be replaced by another value (in case the first task class is 'no-profile', but this is something orthogonal to your problem and to this PR;
 - We don't want the code to loop over all possible values and initialize to -1. jdf2c.c already loops over all entries in the profiling_array, and initializes the entries that are not 'no-profile'. You should just change that part of the code that skips over the 'no-profile' task classes to set the entry to -1 instead of leaving it untouched.

For the latter part, look for `dump_profiling_init` in jdf2c.c. You'll see that there is a loop over all task classes ('functions' in jdf2c.c), and for each task class, we call `dump_profiling_init` using the `UTIL_DUMP_LIST` macro that iterates over `functions`. `dump_profiling_init` checks if profiling is enabled for the task class passed as parameter, and if it is not, it just adds nothing to `info->sa`, otherwise it adds a call to `parsec_profiling_dictionary_add_keyword` to initialize two elements of the `profiling_array` array. You should just need to modify this function (`dump_profiling_init`) to add the initialization of the appropriate two elements of `profiling_array` to -1 instead of skipping over it.

**Comment by DSMishler** [comment]
I checked to make sure the CI check passed and then squashed the commits (on my fork) so that PaRSEC won't have messy and non-compiling commits in its history for the merge.

File: parsec/interfaces/ptg/ptg-compiler/jdf2c.c
Comment (therault) [inline parsec/interfaces/ptg/ptg-compiler/jdf2c.c:897]: Do not leave a commented section like this in your code, if it's useless code, remove it before the merge.

File: parsec/mca/pins/pins.c
Comment (therault) [inline parsec/mca/pins/pins.c:36]: To discuss: it's nice to put this check here, and not in each cb_func, because this removes code duplication, but at the same time, the test below changes the behavior of PINS:

 - if PARSEC_PROF_TRACE is OFF, and task != NULL, no callback is called. This could be less intrusive by protecting the entire block (lines 40 to 47) behind the #ifdef
 - if a PINS module doesn't use parsec_profiling in cb_func, but uses it to take some other statistics, it's not called anyway.

I think we should pay the cost of code duplication here, to keep the feature that some pins module can collect statistics without profiling.

Comment (DSMishler) [reply parsec/mca/pins/pins.c:37]: w.r.t. line 897 in `jdf2c.c`, you're right. Oops, shouldn't have left that in. Fixed.

I'll be checking this email occasionally while off work. Just wanted to let you know I still see this stuff. I added a commit to change this, but I don't know what you mean by each callback function (`cb_func`). Where is this field set? Should we just write a wrapper that checks for the appropriate fields to be non-NULL and _then_ calls the cb_func?

**Review by therault — APPROVED** [review]

**Review by therault — CHANGES_REQUESTED** [review]
There are copy/paste typos in these changes. I will create a PR from this one and fix them so we can merge faster.

**Comment by DSMishler** [comment]
Closing with the merge of PR #594

**Pull request closed without merging**
